### PR TITLE
chore(workspaces): tests for workspace project creation

### PR DESCRIPTION
**Pull request by cdriesler**
NOTE: The tests are passing but we appear to have some gql validation _very_ upset that I removed an input field. Unsure what to do about this, will check in tmr.

## Description & motivation

We already indirectly implemented "workspace project creation" by adding the `workspaceId` param to stream and project creation resolvers in a previous PR. This finishes the job by bringing the change up to latest standards (code quality, style, and tests)

## Changes:

- Removes the possibility to do this through deprecated `streamCreate`. This feels more correct but I can bring it back.
- Adds tests and supporting queries
  - I think workspace projects is distinct enough from the other gql suites, but it's a bit silly in its own file rn with only one test. If the overhead is too much I'm happy to park it in another file for now and break it out when we have more tests for working with workspace projects.

### CODE REVIEW

**Comment by fabis94** [comment]
@cdriesler I've sent out an invite to Apollo Studio, if you think the changes are safe you can mark them as such (remember to do so for all affected operations) and then once the CI reruns its gonna succeed

File: packages/server/assets/workspacesCore/typedefs/workspaces.graphql [hunk 1]
```diff
@@ -95,6 +95,7 @@ type Workspace {
   description: String
   createdAt: DateTime!
   updatedAt: DateTime!
+  logoUrl: String
```
Comment (fabis94) [inline packages/server/assets/workspacesCore/typedefs/workspaces.graphql:98]: The logoUrl param in the WorkspaceCreateInput type doesn't make any sense. Whether we go with storing the base64 data string in the DB or store it in some kind of blob storage, the client is def not the one posting a logo URL to the server - they're gonna post the base64 string or post it as multiform data or whatever. The logo URL comes from the server (if the server creates one at all).

File: packages/server/modules/workspaces/graph/resolvers/workspaces.ts [hunk 2]
```diff
@@ -314,9 +315,18 @@ export = FF_WORKSPACES_MODULE_ENABLED
...
-          // Get projects in workspace
-          throw new WorkspacesNotYetImplementedError()
+        projects: async (parent, args) => {
+          const getWorkspaceProjects = getWorkspaceProjectsFactory({ getStreams })
+          return await getWorkspaceProjects(
+            {
+              workspaceId: parent.id
+            },
+            {
+              limit: args.limit ?? 25,
```
Comment (fabis94) [inline packages/server/modules/workspaces/graph/resolvers/workspaces.ts:325]: `||` is in most cases better than `??` cause it'll use your fallback value not just when the 1st operand is `null` but when its generally falsy in any way, which is usually what you're after. Surely you want the fallback value irregardless of whether cursor is null, or `false` or an empty string or 0?

File: packages/server/modules/workspaces/services/projects.ts [hunk 3]
```diff
@@ -33,3 +35,42 @@ export const queryAllWorkspaceProjectsFactory = ({
...
   }
+
+type GetWorkspaceProjectsArgs = {
+  workspaceId: string
+}
+
+type GetWorkspaceProjectsOptions = {
+  limit: number | null
+  cursor: string | null
+  filter: UserProjectsFilter | null
```
Comment (fabis94) [inline packages/server/modules/workspaces/services/projects.ts:46]: Even if UserProjectsFilter has the same values you want right now, you shouldn't use it and should introduce your own type cause the semantics are different. The types being the same right now doesn't mean they're gonna be the same in the future. In the future the type may get other more user specific fields added to it, and then your GetWorkspaceProjects field is gonna ask for user-specific filtering fields.

File: packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      context: createTestContext({
+        auth: true,
+        userId: testUser.id,
+        token,
+        role: testUser.role,
+        scopes: AllScopes
+      })
+    })
+
+    const { data } = await apollo.execute(CreateWorkspaceDocument, {
```
Comment (fabis94) [inline packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts:57]: Why not use the `createTestWorkspace` helper?

File: packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      const projectName = cryptoRandomString({ length: 6 })
+
+      const createRes = await apollo.execute(CreateWorkspaceProjectDocument, {
+        input: {
+          name: projectName,
+          workspaceId: workspace.id
+        }
+      })
+
+      const getRes = await apollo.execute(GetWorkspaceProjectsDocument, {
```
Comment (fabis94) [inline packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts:79]: This query should have proper tests that also test pagination filtering etc.

**Review by fabis94 — CHANGES_REQUESTED** [review]

File: packages/server/assets/workspacesCore/typedefs/workspaces.graphql (hunk 1 above)
Comment (cdriesler) [reply packages/server/assets/workspacesCore/typedefs/workspaces.graphql:100]: +1, not sure where this came from, done

File: packages/server/modules/workspaces/graph/resolvers/workspaces.ts (hunk 2 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/graph/resolvers/workspaces.ts]: back and forth on this one but will prefer `||` going forwards, this is a good reason

File: packages/server/modules/workspaces/services/projects.ts (hunk 3 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/services/projects.ts]: roger, done

File: packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts (hunk 4 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts]: I am silly and like to duplicate work (done)

File: packages/server/modules/workspaces/services/projects.ts [hunk 6]
```diff
@@ -33,3 +34,44 @@ export const queryAllWorkspaceProjectsFactory = ({
...
+
+type GetWorkspaceProjectsArgs = {
+  workspaceId: string
+}
+
+type GetWorkspaceProjectsOptions = {
+  limit: number | null
+  cursor: string | null
+  filter: {
+    search?: string | null
```
Comment (fabis94) [inline packages/server/modules/workspaces/services/projects.ts:46]: Workspace.projects typedefs still reference `UserProjectsFilter` btw, u also need a new type there

File: packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    id: '',
+    name: 'John Nobody',
+    email: 'john-nobody@example.org',
+    role: Roles.Server.User
+  }
+
+  before(async () => {
+    await beforeEachContext()
+    await Promise.all(
+      [testUser, testNonWorkspaceMemberUser].map((user) => createTestUser(user))
```
Comment (fabis94) [inline packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts:45]: No need to change this, but there's also `createTestUsers` (plural) that would remove the need to do this Promise.all & mapping bit

File: packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+      workspace = testWorkspace
+
+      const workspaceProjects = [
+        { name: 'Workspace Project A', workspaceId: workspace.id },
+        { name: 'Workspace Project B', workspaceId: workspace.id },
+        { name: 'Workspace Project C', workspaceId: workspace.id }
+      ]
+
+      await Promise.all(
```
Comment (fabis94) [inline packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts:118]: This one has a plural version too: `createTestWorkspaces`

File: packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    before(async () => {
+      const testWorkspace: BasicTestWorkspace = {
+        id: '',
+        ownerId: testUser.id,
+        name: 'My Test Workspace'
+      }
+
+      await createTestWorkspace(testWorkspace, testUser)
+
+      workspace = testWorkspace
```
Comment (fabis94) [inline packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts:110]: Here like in the other PR, there's no need to have 2 variables for the same workspace - testWorkspace and workspace. Just move `testWorkspace` outside of `before()` and use that.

**Review by fabis94 — CHANGES_REQUESTED** [review]

File: packages/server/modules/workspaces/services/projects.ts (hunk 6 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/services/projects.ts:46]: oops, just sloppy, fixed

File: packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts (hunk 7 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts]: I need to learn to read, done

File: packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts (hunk 9 above)
Comment (cdriesler) [reply packages/server/modules/workspaces/tests/integration/projects.graph.spec.ts]: much better, no sure why I did it this way

**Review by fabis94 — APPROVED** [review]

**Pull request merged**